File: Battelle.EPA.WideAreaDecon.Client/src/implementations/parameter/list/sumFraction.ts
import { JsonProperty } from 'typescript-json-serializer';
import ParameterType from '@/enums/parameter/parameterTypes';
import IParameter from '@/interfaces/parameter/IParameter';
import KeyValuePair from './keyValuePair';
import ParameterMetaData from '../ParameterMetaData';

export default class SumFraction implements IParameter {
  @JsonProperty()
  name: string;

  @JsonProperty()
  type: ParameterType = ParameterType.sumFraction;

  @JsonProperty()
  metaData: ParameterMetaData;

  @JsonProperty({
    predicate: () => {
      return KeyValuePair;
    },
  })
  entries: Array<KeyValuePair>;

  public isSet(): boolean {
    return this.entries.every((e) => e.value !== null);
  }

  constructor(name = 'unknown', metaData = new ParameterMetaData(), entries?: Array<KeyValuePair>) {
    this.name = name;
    this.entries = entries !== undefined ? entries : [];
    this.metaData = metaData;
  }

  isEquivalent(other: IParameter): boolean {
    return this.compareValues(other as SumFraction);
  }

  compareValues(other?: SumFraction): boolean {
    if (other === undefined || this.type !== other.type) {
      return false;
    }
    return this.entries.every((kvp: KeyValuePair, i: number) => kvp.isEquivalent(other.entries[i]));
  }
}

File: Battelle.EPA.WideAreaDecon.API/Models/Parameter/List/SumFraction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Xml.Linq;
using System.Threading.Tasks;
using System.Runtime.Serialization;
using Battelle.EPA.WideAreaDecon.API.Enumeration.Parameter;
using Battelle.EPA.WideAreaDecon.API.Interfaces.Parameter;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using NPOI.SS.UserModel;

namespace Battelle.EPA.WideAreaDecon.API.Models.Parameter.List
{
    /// <summary>
    /// Implementation of the sum fraction
    /// </summary>
    public class SumFraction : IParameter
    {
        private static int NameLocation => 2;

        public string Name { get; set; }

        [JsonConverter(typeof(StringEnumConverter))]
        public ParameterType Type => ParameterType.SumFraction;

        public KeyValuePair[] Entries { get; set; }

        public ParameterMetaData MetaData { get; set; }

        public static SumFraction FromExcel(IRow information)
        {
            return new SumFraction()
            {
                Name = information.GetCell(NameLocation)?.ToString() ?? throw new SerializationException("Parameter has no name associated with it in Excel"),
                MetaData = ParameterMetaData.FromExcel(information)
            };
        }
    }
}
